### ISSUE TITLE
Azure all_valid_ems_in_zone does an Azure API call every sync_workers loop

### ISSUE THREAD

**Issue by agrare** [ManageIQ/manageiq-providers-azure#438]
The Azure Provider only starts event_catchers if the `Microsoft.Insights` resource provider is registered.  This is done by calling [`insights?`](https://github.com/ManageIQ/manageiq-providers-azure/blob/master/app/models/manageiq/providers/azure/cloud_manager.rb#L42-L48) which is called by the [`supports?(:timeline)`](https://github.com/ManageIQ/manageiq-providers-azure/blob/master/app/models/manageiq/providers/azure/cloud_manager.rb#L28-L32) supports feature block.

The problem arises when the `Azure::CloudManager::EventCatcher` overrides [`all_valid_ems_in_zone`](https://github.com/ManageIQ/manageiq-providers-azure/blob/master/app/models/manageiq/providers/azure/cloud_manager/event_catcher.rb#L7-L13) to check `supports?(:timeline)`.

This looks innocuous but it is actually doing an Azure API call every time `Azure::CloudManager::EventCatcher.all_valid_ems_in_zone` is called.  This is done by the [`PerEmsWorkerMixin.sync_workers`](https://github.com/ManageIQ/manageiq/blob/master/app/models/mixins/per_ems_worker_mixin.rb#L32) method (by way of [`desired_queue_names`](https://github.com/ManageIQ/manageiq/blob/master/app/models/mixins/per_ems_worker_mixin.rb#L25-L27)).

Long story short we're doing an Azure API call from the MiqServer to check this feature every 15 seconds.

**Comment by agrare** [comment]
https://github.com/ManageIQ/manageiq/issues/21061

**Comment by Fryguy** [comment]
As a tactical fix, if we can't get all the capabilities work in time, I recommend using a cache_with_timeout of maybe 1.hour

**Comment by jrafanie** [comment]
From chat with @agrare:

`insights?` is referenced as being a problem but, I'm finding that the time is pretty bad just connecting to the provider...

For example, here's what `insights?` looks like in the environment that opened this issue:

```
irb(main):003:0> ManageIQ::Providers::Azure::CloudManager.first.insights?
{"@timestamp":"2021-02-17T19:55:59.751397 ","hostname":"orchestrator-fb9c5b75b-4wd4p","pid":63286,"tid":"2b1d439df964","level":"info","message":"MIQ(ManageIQ::Providers::Azure::CloudManager#with_provider_connection) Connecting through ManageIQ::Providers::Azure::CloudManager: [Azure_East_US]"}
RestClient.post "https://login.microsoftonline.com/<GUID>/oauth2/token", "grant_type=client_credentials&client_id=<ID_STRING>&client_secret=<SECRET>&resource=https%3A%2F%2Fmanagement.core.windows.net%2F", "Accept"=>"*/*", "Content-Length"=>"181", "Content-Type"=>"application/x-www-form-urlencoded", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 1637 bytes, 0.59s
RestClient.get "https://management.azure.com/subscriptions?api-version=2016-06-01", "Accept"=>"application/json", "Authorization"=>"Bearer eyJ0eXAi...", "Content-Type"=>"application/json", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 342 bytes, 0.39s
RestClient.get "https://management.azure.com/subscriptions/<SUBSCRIPTION_ID>/providers?api-version=2015-01-01", "Accept"=>"application/json", "Authorization"=>"Bearer eyJ0eXAi...", "Content-Type"=>"application/json", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 933353 bytes, 2.52s
RestClient.get "https://management.azure.com/subscriptions/<SUBSCRIPTION_ID>/providers/Microsoft.Insights?api-version=2019-05-01", "Accept"=>"application/json", "Authorization"=>"Bearer eyJ0eXAi...", "Content-Type"=>"application/json", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 22712 bytes, 0.30s
=> true
```

But, just connecting is bad:

```
irb(main):003:0> cm.with_provider_connection do |conf|
irb(main):004:1* puts "1"
irb(main):005:1> end
{"@timestamp":"2021-02-17T20:06:40.655390 ","hostname":"orchestrator-fb9c5b75b-4wd4p","pid":63888,"tid":"2afd262a5960","level":"info","message":"MIQ(ManageIQ::Providers::Azure::CloudManager#with_provider_connection) Connecting through ManageIQ::Providers::Azure::CloudManager: [Azure_East_US]"}
RestClient.post "https://login.microsoftonline.com/<GUID>/oauth2/token", "grant_type=client_credentials&client_id=<ID_STRING>&client_secret=<SECRET>&resource=https%3A%2F%2Fmanagement.core.windows.net%2F", "Accept"=>"*/*", "Content-Length"=>"181", "Content-Type"=>"application/x-www-form-urlencoded", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 1637 bytes, 0.33s
RestClient.get "https://management.azure.com/subscriptions?api-version=2016-06-01", "Accept"=>"application/json", "Authorization"=>"Bearer eyJ0eXAi...", "Content-Type"=>"application/json", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 342 bytes, 0.43s
RestClient.get "https://management.azure.com/subscriptions/<SUBSCRIPTION_ID>/providers?api-version=2015-01-01", "Accept"=>"application/json", "Authorization"=>"Bearer eyJ0eXAi...", "Content-Type"=>"application/json", "User-Agent"=>"rest-client/2.1.0 (linux-gnu x86_64) ruby/2.6.3p62"
# => 200 OK | application/json 933103 bytes, 2.60s
1
```

It turns out, raw_connect creates a [Azure::Armrest::Configuration passing a subscription_id in the options](https://github.com/ManageIQ/manageiq-providers-azure/blob/44b3c8e8b62d5cd45f21abe6a4a6a0afd6c38647/app/models/manageiq/providers/azure/manager_mixin.rb#L161).

In the initialize, it [calls the setter on all remaining options passed](https://github.com/ManageIQ/azure-armrest/blob/3a9ff6ce732a517293ee27c0c18ad985c09c77a2/lib/azure/armrest/configuration.rb#L133).


That calls [fetch_providers](https://github.com/ManageIQ/azure-armrest/blob/3a9ff6ce732a517293ee27c0c18ad985c09c77a2/lib/azure/armrest/configuration.rb#L159), which is the slow 2.6 second request.  That result is saved in the `@providers` variable as a list.  @agrare noted that we call `::Azure::Armrest::ResourceProviderService.new`  in the [insights?](https://github.com/ManageIQ/manageiq-providers-azure/blob/master/app/models/manageiq/providers/azure/cloud_manager.rb#L42-L48) method too but uses `.get`. 

If we lazy load the providers and store them in a format we can reuse, it should be faster to connect and less redundant requests.

**Comment by Fryguy** [comment]
@agrare On an upgrade, this value is going to default to nil, so until they re-verify, they won't have a valid value.  Is that a problem?

**Comment by agrare** [comment]
We do an auth check every hour and I think we do one on startup but I'll have to verify that

**Comment by Fryguy** [comment]
If we do one at startup then we are good.

**Comment by agrare** [comment]
Worker startup actually checks it, I cleared the `ems.capabilities` and started the azure event_catcher and it checked `Microsoft.Insights` and started up properly.
